File: app/src/main/java/com/example/epigram/MyAdapterArticles.java
package com.example.epigram;

import android.app.Activity;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.bumptech.glide.Glide;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.MultiTransformation;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.load.resource.bitmap.CenterCrop;
import com.bumptech.glide.load.resource.bitmap.RoundedCorners;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.RequestOptions;
import com.bumptech.glide.request.target.Target;
import com.example.epigram.data.Post;
import com.jakewharton.rxbinding2.view.RxView;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import kotlin.random.Random;
import org.joda.time.DateTime;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.TimeUnit;

public class MyAdapterArticles extends RecyclerView.Adapter<MyAdapterArticles.MyViewHolder> {

    public static int SEARCH_PAGE_INDEX = 100;
    public static int HOME_PAGE_INDEX = 0;

    private Post vcars = new Post("advert", "advert-vcars", "Get 20% off vcars as a student: download the app now", "", "https://www.v-cars.com/wp-content/uploads/2019/05/VCarsLogo-Resized.png", "ADVERT", Arrays.asList("ADVERT"), DateTime.now(), "");
    private boolean advertTime = true;

    public List<Post> posts = new ArrayList<>();
    private final MultiTransformation<Bitmap> multiTransformation;
    private LoadNextPage loadNextPage = null;
    private int resultTotal = 0;
    private Context context;

    private int pageIndex; // 100 for search recycler view

    public void clear() {
        posts.clear();
        notifyDataSetChanged();
    }

    public void setPostList(List<Post> checkSame) {

        DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(new DiffUtil.Callback() {
            @Override
            public int getOldListSize() {
                return posts.size();
            }

            @Override
            public int getNewListSize() {
                return checkSame.size();
            }

            @Override
            public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
                return posts.get(oldItemPosition) == checkSame.get(newItemPosition);
            }

            @Override
            public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
                return posts.get(oldItemPosition).getId().equals(checkSame.get(newItemPosition).getId());
            }
        });
        posts = checkSame;
        diffResult.dispatchUpdatesTo(this);
    }

    public class MyViewHolder extends RecyclerView.ViewHolder{

        public TextView title;
        public ImageView titleImage;
        public LinearLayout linearLayout;
        //public TextView tag;
        public RecyclerView tags;
        // public TextView date; // for date on top of image
        public TextView dateAlt;
        //public TextView sectionTitle;
        public TextView searchResults;
        public boolean imageLoaded = false;

        private Disposable disposable = null;

        public MyViewHolder(LinearLayout l){
            super(l);
            if(pageIndex == SEARCH_PAGE_INDEX){
                searchResults = l.findViewById(R.id.search_results_number);
            }
            title = l.findViewById(R.id.post_title);
            //tag = l.findViewById(R.id.tag_text);
            tags = l.findViewById(R.id.recycler_view_tag);
            // date = l.findViewById(R.id.post_date); // for top of image
            dateAlt = l.findViewById(R.id.post_date_alternate);
            titleImage = l.findViewById(R.id.post_image);
            this.linearLayout = l;
            //sectionTitle = l.findViewById(R.id.section_text);
        }
    }

    @Override
    public void onViewDetachedFromWindow(@NonNull MyViewHolder holder) {
        super.onViewDetachedFromWindow(holder);
        if(holder.disposable != null) {
            holder.disposable.dispose();
        }
    }

    @Override
    public void onViewAttachedToWindow(@NonNull MyViewHolder holder) {
        super.onViewAttachedToWindow(holder);

        if(holder.disposable != null) {
            holder.disposable.dispose();
        }
        holder.disposable = RxView.clicks(holder.linearLayout).throttleFirst(500, TimeUnit.MILLISECONDS).observeOn(AndroidSchedulers.mainThread()).subscribe(empty -> {
            holder.titleImage.setTransitionName("article_header");

            loadNextPage.onPostClicked(posts.get(holder.getAdapterPosition()), holder.imageLoaded?holder.titleImage:null);
        });
    }

    public MyAdapterArticles(Context context, List<Post> posts, LoadNextPage loadNext, int position) {
        this.context = context;
        this.posts = posts;
        multiTransformation = new MultiTransformation<>(new CenterCrop(),new RoundedCorners(40));
        loadNextPage = loadNext;
        pageIndex = position;
    }

    public void addPosts(List<Post> postsNew){


        List<Post> checkSame = new ArrayList<>(posts);

        checkSame.addAll(postsNew);

        DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(new DiffUtil.Callback() {
            @Override
            public int getOldListSize() {
                return posts.size();
            }

            @Override
            public int getNewListSize() {
                return checkSame.size();
            }

            @Override
            public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
                return posts.get(oldItemPosition) == checkSame.get(newItemPosition);
            }

            @Override
            public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
                return posts.get(oldItemPosition).getId().equals(checkSame.get(newItemPosition).getId());
            }
        });

        if(advertTime) {
            checkSame.add(vcars);
        }
        advertTime = !advertTime;

        posts = checkSame;

        if(posts.get(0).getDate().plusWeeks(1).isBeforeNow()) posts.remove(0);



        diffResult.dispatchUpdatesTo(this);
        //posts = new ListUtils().duplicatePost(new ArrayList<>(posts));
        //notifyDataSetChanged();
    }

    @Override
    public MyAdapterArticles.MyViewHolder onCreateViewHolder(ViewGroup parent, int viewType){
        if(pageIndex == HOME_PAGE_INDEX && viewType == 1){
            if(posts.get(0).getDate().plusWeeks(1).isBeforeNow() && posts.get(0).getTags().contains("breaking-news")){ //
                posts.remove(0);
                LinearLayout l = (LinearLayout) LayoutInflater.from(parent.getContext()).inflate(R.layout.element_news_article_first, parent, false);
                MyViewHolder vh = new MyViewHolder(l);
                return vh;

            }
            else{
                //if(posts.get(0).getTags().contains("breaking-news")) posts.remove(0);
                LinearLayout l = (LinearLayout) LayoutInflater.from(parent.getContext()).inflate(R.layout.element_news_article_breaking, parent, false);
                MyViewHolder vh = new MyViewHolder(l);
                return vh;
            }
        }
        else if(viewType == 1 && pageIndex == SEARCH_PAGE_INDEX){
            LinearLayout l = (LinearLayout) LayoutInflater.from(parent.getContext()).inflate(R.layout.element_news_article_first, parent, false);
            MyViewHolder vh = new MyViewHolder(l);
            return vh;
        }
        else {
            LinearLayout l = (LinearLayout) LayoutInflater.from(parent.getContext()).inflate(R.layout.element_news_article, parent, false);
            MyViewHolder vh = new MyViewHolder(l);
            return vh;
        }
    }

    @Override
    public int getItemViewType(int position) {
        if(position == 0 && (pageIndex == HOME_PAGE_INDEX || pageIndex == SEARCH_PAGE_INDEX)) return 1;
        else return 2;
    }

    @Override
    public void onBindViewHolder(MyViewHolder holder, int position){
        holder.tags.setLayoutManager(new LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false));
        holder.tags.setItemAnimator(new DefaultItemAnimator());
        holder.tags.setAdapter(new MyAdapterTag(posts.get(position).getTags()));
        setPosts(holder, position);
    }

    public void setPosts(MyViewHolder holder, int position){
        if(pageIndex == SEARCH_PAGE_INDEX && position == 0){
            holder.searchResults.setText(holder.searchResults.getResources().getQuantityString(R.plurals.results, resultTotal,resultTotal));//Integer.toString(posts.size()));
        }
        List<String> tag = posts.get(position).getTags();
        tag.removeAll(Arrays.asList("featured top", "carousel", "one sidebar"));
        holder.title.setText((posts.get(position).getTitle()));
        //holder.tag.setText(tag.get(0).toUpperCase());//posts.get(position).getTag());
        holder.dateAlt.setText(posts.get(position).getDate().toString("MMM d, yyyy"));

        Glide.with(holder.titleImage).load(posts.get(position).getImage()).placeholder(R.drawable.placeholder_background).apply(RequestOptions.bitmapTransform(multiTransformation)).listener(new RequestListener<Drawable>() {
            @Override
            public boolean onLoadFailed(@Nullable GlideException e, Object model, Target<Drawable> target, boolean isFirstResource) {
                holder.imageLoaded = false;
                return false;
            }

            @Override
            public boolean onResourceReady(Drawable resource, Object model, Target<Drawable> target, DataSource dataSource, boolean isFirstResource) {
                holder.imageLoaded = true;
                return false;
            }
        }).into(holder.titleImage);
        if(position > getItemCount() - 2) loadNextPage.bottomReached();
    }

    @Override
    public int getItemCount(){
        return posts.size();
    }

    public interface LoadNextPage{
        public void bottomReached();

        public void onPostClicked(Post clicked, ImageView titleImage);
    }

    public void setResultTotal(int total){
        resultTotal = total;
        notifyItemChanged(0);
    }

}

File: app/src/main/java/com/example/epigram/MyAdapterArticles.kt
package com.example.epigram

import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.MultiTransformation
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.load.resource.bitmap.CenterCrop
import com.bumptech.glide.load.resource.bitmap.RoundedCorners
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.RequestOptions
import com.bumptech.glide.request.target.Target
import com.example.epigram.data.Post
import com.jakewharton.rxbinding2.view.RxView
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable

import java.util.ArrayList
import java.util.concurrent.TimeUnit

class MyAdapterArticles(
    posts: MutableList<Post>, loadNext: LoadNextPage, private val pageIndex: Int // 100 for search recycler view
) : RecyclerView.Adapter<MyAdapterArticles.MyViewHolder>() {

    var posts: MutableList<Post> = ArrayList<Post>()
    private val multiTransformation: MultiTransformation<Bitmap>
    private var loadNextPage: LoadNextPage? = null

    private var resultTotal = 0

    fun clear() {
        posts.clear()
        notifyDataSetChanged()
    }

    fun setPostList(checkSame: MutableList<Post>) {

        val diffResult = DiffUtil.calculateDiff(object : DiffUtil.Callback() {
            override fun getOldListSize(): Int {
                return posts.size
            }

            override fun getNewListSize(): Int {
                return checkSame.size
            }

            override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                return posts[oldItemPosition] === checkSame[newItemPosition]
            }

            override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                return posts[oldItemPosition].getId().equals(checkSame[newItemPosition].getId())
            }
        })
        posts = checkSame
        diffResult.dispatchUpdatesTo(this)
    }

    inner class MyViewHolder(var linearLayout: LinearLayout) : RecyclerView.ViewHolder(linearLayout) {

        var title: TextView
        var titleImage: ImageView
        var tag: TextView
        // public TextView date; // for date on top of image
        var dateAlt: TextView
        //public TextView sectionTitle;
        var searchResults: TextView

        var imageLoaded = false

        private val disposable: Disposable? = null

        init {
            if (pageIndex == SEARCH_PAGE_INDEX) {
                searchResults = linearLayout.findViewById(R.id.search_results_number)
            }
            title = linearLayout.findViewById(R.id.post_title)
            tag = linearLayout.findViewById(R.id.tag_text)
            // date = l.findViewById(R.id.post_date); // for top of image
            dateAlt = linearLayout.findViewById(R.id.post_date_alternate)
            titleImage = linearLayout.findViewById(R.id.post_image)
            //sectionTitle = l.findViewById(R.id.section_text);
        }
    }

    override fun onViewDetachedFromWindow(holder: MyViewHolder) {
        super.onViewDetachedFromWindow(holder)
        if (holder.disposable != null) {
            holder.disposable.dispose()
        }
    }

    override fun onViewAttachedToWindow(holder: MyViewHolder) {
        super.onViewAttachedToWindow(holder)

        if (holder.disposable != null) {
            holder.disposable.dispose()
        }
        holder.disposable = RxView.clicks(holder.linearLayout).throttleFirst(500, TimeUnit.MILLISECONDS)
            .observeOn(AndroidSchedulers.mainThread()).subscribe { empty ->
            holder.titleImage.transitionName = "article_header"

            loadNextPage!!.onPostClicked(
                posts[holder.adapterPosition],
                if (holder.imageLoaded) holder.titleImage else null
            )
        }
    }

    init {
        this.posts = posts
        multiTransformation = MultiTransformation(CenterCrop(), RoundedCorners(32))
        loadNextPage = loadNext
    }

    fun addPosts(postsNew: List<Post>) {

        val checkSame = ArrayList(posts)
        checkSame.addAll(postsNew)

        val diffResult = DiffUtil.calculateDiff(object : DiffUtil.Callback() {
            override fun getOldListSize(): Int {
                return posts.size
            }

            override fun getNewListSize(): Int {
                return checkSame.size
            }

            override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                return posts[oldItemPosition] === checkSame.get(newItemPosition)
            }

            override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                return posts[oldItemPosition].getId().equals(checkSame.get(newItemPosition).getId())
            }
        })
        posts = checkSame

        diffResult.dispatchUpdatesTo(this)
        //posts = new ListUtils().duplicatePost(new ArrayList<>(posts));
        //notifyDataSetChanged();
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyAdapterArticles.MyViewHolder {
        if (viewType == 1 && pageIndex == HOME_PAGE_INDEX) {
            val l = LayoutInflater.from(parent.context).inflate(
                R.layout.element_news_article_breaking,
                parent,
                false
            ) as LinearLayout
            return MyViewHolder(l)
        } else if (viewType == 1 && pageIndex == SEARCH_PAGE_INDEX) {
            val l = LayoutInflater.from(parent.context).inflate(
                R.layout.element_news_article_first,
                parent,
                false
            ) as LinearLayout
            return MyViewHolder(l)
        } else {
            val l = LayoutInflater.from(parent.context).inflate(
                R.layout.element_news_article,
                parent,
                false
            ) as LinearLayout
            return MyViewHolder(l)
        }
    }

    override fun getItemViewType(position: Int): Int {
        return if (position == 0 && (pageIndex == HOME_PAGE_INDEX || pageIndex == SEARCH_PAGE_INDEX))
            1
        else
            2
    }

    override fun onBindViewHolder(holder: MyViewHolder, position: Int) {
        setPosts(holder, position)
    }

    fun setPosts(holder: MyViewHolder, position: Int) {
        if (pageIndex == SEARCH_PAGE_INDEX && position == 0) {
            holder.searchResults.text =
                holder.searchResults.resources.getQuantityString(
                    R.plurals.results,
                    resultTotal,
                    resultTotal
                )//Integer.toString(posts.size()));
        }
        holder.title.setText(posts[position].getTitle())
        holder.tag.setText(posts[position].getTag())
        //holder.date.setText(posts.get(position).getDate().toString("MMM d, yyyy")); // date for on top of image
        holder.dateAlt.setText(posts[position].getDate().toString("MMM d, yyyy"))

        Glide.with(holder.titleImage).load(posts[position].getImage()).placeholder(R.drawable.placeholder_background)
            .apply(RequestOptions.bitmapTransform(multiTransformation)).listener(object : RequestListener<Drawable> {
            override fun onLoadFailed(
                e: GlideException?,
                model: Any,
                target: Target<Drawable>,
                isFirstResource: Boolean
            ): Boolean {
                holder.imageLoaded = false
                return false
            }

            override fun onResourceReady(
                resource: Drawable,
                model: Any,
                target: Target<Drawable>,
                dataSource: DataSource,
                isFirstResource: Boolean
            ): Boolean {
                holder.imageLoaded = true
                return false
            }
        }).into(holder.titleImage)
        if (position > itemCount - 2) loadNextPage!!.bottomReached()
    }

    override fun getItemCount(): Int {
        return posts.size
    }

    interface LoadNextPage {
        fun bottomReached()

        fun onPostClicked(clicked: Post, titleImage: ImageView?)
    }

    fun setResultTotal(total: Int) {
        resultTotal = total
        notifyItemChanged(0)
    }

    companion object {

        var SEARCH_PAGE_INDEX = 100
        var HOME_PAGE_INDEX = 0
    }

}


